### ISSUE TITLE
Support for use_datetime in twisted.web.xmlrpc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3219]
|<img alt="chadmaine's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| chadmaine reported|
|-|-|
|Trac ID|trac#3219|
|Type|enhancement|
|Created|2008-05-01 15:31:47Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-datetime-3219|

Python's xmlrpclib.loads will convert xmlrpclib.DateTime objects to datetime.datetime objects when the use_datetime arg is set to True.

I've attached a simple diff that provides this functionality for twisted.web.xmlrpc.

Attachments:

* [xmlrpc_datetime.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2dd/2dd7939d675f664ebc671a9c7c09a2fea1e2dc8b/8b0456af88e01cfbbdfe6caaccf60118ccb23bd6.diff) (3413 bytes) - added by chadmaine on 2008-05-01 15:37:23Z - 
* [xmlrpc_datetime_r28820.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2dd/2dd7939d675f664ebc671a9c7c09a2fea1e2dc8b/09843028bcd80ee9a8f9319219864c6f93fa7db8.diff) (3085 bytes) - added by chadmaine on 2010-04-15 22:21:59Z - Updated against r28820
* [xmlrpc_datetime_r28853.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2dd/2dd7939d675f664ebc671a9c7c09a2fea1e2dc8b/3d20526d98880d6d72ef7ebb982a38b281d5e27e.diff) (13938 bytes) - added by chadmaine on 2010-04-23 20:28:35Z - Updated diff with exarkun recommended changes, please review

<details><summary>Searchable metadata</summary>

```
trac-id__3219 3219
type__enhancement enhancement
reporter__chadmaine chadmaine
priority__normal normal
milestone__ 
branch__branches_xmlrpc_datetime_3219 branches/xmlrpc-datetime-3219
branch_author__chadmaine__exarkun chadmaine, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__xmlrpc_datetime xmlrpc datetime
time__1209655907000000 1209655907000000
changetime__1272637770000000 1272637770000000
version__None None
owner__ 
cc__khorn
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="chadmaine's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|chadmaine commented|
|-|-|

Anyone had a chance to look at this patch?  We continue to patch Twisted manually to use this feature.

Chad

**Comment by twisted-trac** [comment]
|<img alt="rion's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|rion commented|
|-|-|

i also want this feature =)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_4"></a>|@khorn commented|
|-|-|

This change seems reasonable.  Does anyone else see a reason not to proceed with this?  Will it conflict with anything else?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to chadmaine|
|-|-|

Hi Chad, thanks for the update.  The patch looks pretty good, but there are a couple changes it'll need before we can include it in trunk:

  1. The new `useDateTime` attribute/parameter of `XMLRPC` and `Proxy` needs to be documented.  Convention is to document it as an ivar on the class (rather than as a parameter to `__init__`).  This should have been done for `allowNone` as well, by whoever added that.
  1. The two new features (server-side and client-side datetime support) need unit tests.  You can find the existing unit tests for this module in `twisted/web/test/test_xmlrpc.py`.
  1. `use_datetime` was introduced in Python 2.5.  Twisted still supports Python 2.4, so we need to do something about that case.  Certainly we should not pass it to xmlrpclib if we're running on Python 2.4, as that will raise an exception.  Possibly we should also raise an exception in `XMLRPC` and `Proxy` if we're running on Python 2.4 and the application passes `useDateTime` in, to mark this error as early as possible.
  1. If you like, it'd also be helpful to include a news fragment for this change: http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

Thanks again for the patch.  To make sure the issue gets attention again quickly when you attach a new patch, please add the "review" keyword to it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28854]) Branching to 'xmlrpc-datetime-3219'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Thanks for the updated patch!  Some comments about it:

  1. `XMLRPCAllowNoneTestCase.setUp` should have a docstring.
  1. You can use `TestCase.addCleanup` in `XMLRPCAllowNoneTestCase.setUp` to get rid of the listening port, instead of using `tearDown`.  It doesn't make much difference in this case, but it does mean you can completely get rid of `tearDown` (which means no one needs to write a docstring for it :).
  1. Test method docstrings shouldn't start with "Test that ...".  They can just describe the behavior being tested.
  1. The `allowNone` tests and the `useDateTime` tests could probably be factored into a mixin pretty easily, reducing the total amount of code needed to test these two features.
  1. The new tests need to skip if being run on Python 2.4, since they'll fail with a `RuntimeError`, but that's expected and correct.
  1. There should also be tests for the `RuntimeError` raised if this code is running on Python 2.4.
  1. Since `XMLRPC` is a classic class, property setters don't work on it.  The check needs to be done in `__setattr__`.
  1. There's a mistake in the new docs for `useDateTime` - they say "<= 2.5" but they mean ">= 2.5".
  1. Most of the `_QueryFactory` docstring changes look good, but a few of the `__init__` parameters don't actually end up as instance attributes, so these should stay as docs in `__init__` (sorry if I wasn't clear about that).
  1. `_QueryFactory` is private, so it probably doesn't need to be protected against a bad value for `useDateTime`; the checks in `XMLRPC` will make sure a bad value isn't passed in.

I'm going to make these changes and re-submit for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Made the above changes, someone else please review.

**Comment by twisted-trac** [comment]
|<img alt="chadmaine's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|chadmaine commented|
|-|-|

Replying to [exarkun](#note_11):
> Made the above changes, someone else please review.

Looking it over, it looks good.  Thanks for your help, exarkun.  I did notice the docs still have my 'Python <= 2.5' in them in a couple of places.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good, other than same comment as chadmaine (you need to replace Python <= 2.5 with >= 2.5 in a few places.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [28880]) Merge xmlrpc-datetime-3219

Author: chadmaine, exarkun
Reviewer: exarkun, chadmaine, itamar
Fixes: [#3219](https://github.com/twisted/twisted/issues/3219)

Add support for representing timestamps received over XML-RPC as
datetime instances, on versions of Python where xmlrpclib supports
that.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
